### ISSUE TITLE
SharedExampleContext: cast of object to protocol type always nil

### ISSUE THREAD

**Issue by devdrey** [Quick/Quick#653]
- [x] I have read [CONTRIBUTING](https://github.com/Quick/Quick/blob/master/CONTRIBUTING.md) and have done my best to follow them.

### What did you do?

I tried to create a test for a default implementation of an protocol as `sharedExample` to use also to test the default behavior of this protocol also in other implementations of that protocol.

The protocol has a default implementation of a method which I wanted to test in the `sharedExample` therefore I needed to pass an implementation of that protocol within a `sharedExampleContext` and cast it to the protocol type.

`let someImplementation = sharedExampleContext()["instance"] as? MyProtocol`

### What did you expect to happen?

I expected `someImplementation` to become non `nil` to call the a method on it to test within the `sharedExample`

### What actually happened instead?

`someImplementation` always becomes `nil`

### Environment

List the software versions you're using:

 - Quick: *1.0.0*
 - Nimble: *5.0.0*
 - Xcode Version: *.1 (8B62)*
 - Swift Version: *3.0.1*
 - Cocoapods: *1.1.1*

### Example
[SwiftPlayground.playground.zip](https://github.com/Quick/Quick/files/629813/SwiftPlayground.playground.zip)

The reason of this behavior is the return type of the `SharedExampleContext` that is `NSDictionary`.

I created an example in a Playground to test and verify that behavior with the `NSDictionary` type.

```
import Foundation

protocol AProtocol {
    func test() -> String
}

extension AProtocol {
    func test() -> String {
        return "string"
    }
}

struct AImplementation: AProtocol {
    // will also inherit the default implementation of test()
}

let aClosure: ( () -> NSDictionary ) = { // mimics Quicks SharedExampleContext
    ["A implementation": AImplementation()]
}

let aImplementation = aClosure()["A implementation"] as? AProtocol // always nil

aImplementation?.test() // also nil
```

This example behaves exactly the same as the `sharedExample` in quick.

There are two options I found to solve this.
1. Replacing the `NSDictionary` with `[String: Any]`
2. Defining `AImplementation ` as `class` instead of `struct`

In order to write a shared test for the default implementation of a protocol I thing Quick should use the Swift dictionary type rather than the objc type.

**Comment by istx25** [comment]
Thanks for reaching out, @devdrey. We'll do our best to resolve this issue and answer your question as soon as possible. I'm a little busy right now so there may be a bit of delay unless someone else can answer the question in the meantime.

**Comment by jeffh** [comment]
Fixed in #654. Closing.

**Comment by esad** [comment]
Note: there's a bug in swift (https://bugs.swift.org/browse/SR-3871) that causes downcasts from optionals (typed as Any in the context dictionary) to protocols to always fail (and yield nil).
